File: clients/php/Codegame/ClientMessage.php
<?php

namespace Codegame {
    require_once 'Debugging/DebugCommand.php';
    require_once 'Model/Order.php';
    require_once 'Stream.php';

    /**
     * Message sent from client
     */
    abstract class ClientMessage
    {
        /**
         * Write ClientMessage to output stream
         */
        abstract function writeTo(\OutputStream $stream): void;

        /**
         * Read ClientMessage from input stream
         */
        static function readFrom(\InputStream $stream): ClientMessage
        {
            $tag = $stream->readInt32();
            if ($tag == \Codegame\ClientMessage\DebugMessage::TAG) {
                return \Codegame\ClientMessage\DebugMessage::readFrom($stream);
            }
            if ($tag == \Codegame\ClientMessage\OrderMessage::TAG) {
                return \Codegame\ClientMessage\OrderMessage::readFrom($stream);
            }
            if ($tag == \Codegame\ClientMessage\DebugUpdateDone::TAG) {
                return \Codegame\ClientMessage\DebugUpdateDone::readFrom($stream);
            }
            if ($tag == \Codegame\ClientMessage\RequestDebugState::TAG) {
                return \Codegame\ClientMessage\RequestDebugState::readFrom($stream);
            }
            throw new Exception('Unexpected tag value');
        }
    }
}

namespace Codegame\ClientMessage {
    /**
     * Ask app to perform new debug command
     */
    class DebugMessage extends \Codegame\ClientMessage
    {
        const TAG = 0;
    
        /**
         * Command to perform
         */
        public \Debugging\DebugCommand $command;
    
        function __construct(\Debugging\DebugCommand $command)
        {
            $this->command = $command;
        }
    
        /**
         * Read DebugMessage from input stream
         */
        public static function readFrom(\InputStream $stream): DebugMessage
        {
            $command = \Debugging\DebugCommand::readFrom($stream);
            return new DebugMessage($command);
        }
        
        /**
         * Write DebugMessage to output stream
         */
        public function writeTo(\OutputStream $stream): void
        {
            $stream->writeInt32(DebugMessage::TAG);
            $this->command->writeTo($stream);
        }
    }

    /**
     * Reply for ServerMessage::GetOrder
     */
    class OrderMessage extends \Codegame\ClientMessage
    {
        const TAG = 1;
    
        /**
         * Player's order
         */
        public \Model\Order $order;
    
        function __construct(\Model\Order $order)
        {
            $this->order = $order;
        }
    
        /**
         * Read OrderMessage from input stream
         */
        public static function readFrom(\InputStream $stream): OrderMessage
        {
            $order = \Model\Order::readFrom($stream);
            return new OrderMessage($order);
        }
        
        /**
         * Write OrderMessage to output stream
         */
        public function writeTo(\OutputStream $stream): void
        {
            $stream->writeInt32(OrderMessage::TAG);
            $this->order->writeTo($stream);
        }
    }

    /**
     * Signifies finish of the debug update
     */
    class DebugUpdateDone extends \Codegame\ClientMessage
    {
        const TAG = 2;
    
    
        function __construct()
        {
        }
    
        /**
         * Read DebugUpdateDone from input stream
         */
        public static function readFrom(\InputStream $stream): DebugUpdateDone
        {
            return new DebugUpdateDone();
        }
        
        /**
         * Write DebugUpdateDone to output stream
         */
        public function writeTo(\OutputStream $stream): void
        {
            $stream->writeInt32(DebugUpdateDone::TAG);
        }
    }

    /**
     * Request debug state from the app
     */
    class RequestDebugState extends \Codegame\ClientMessage
    {
        const TAG = 3;
    
    
        function __construct()
        {
        }
    
        /**
         * Read RequestDebugState from input stream
         */
        public static function readFrom(\InputStream $stream): RequestDebugState
        {
            return new RequestDebugState();
        }
        
        /**
         * Write RequestDebugState to output stream
         */
        public function writeTo(\OutputStream $stream): void
        {
            $stream->writeInt32(RequestDebugState::TAG);
        }
    }
}

File: clients/rust/src/codegame/client_message.rs
use super::*;

/// Message sent from client
#[derive(Clone, Debug)]
pub enum ClientMessage {
    /// Ask app to perform new debug command
    DebugMessage {
        /// Command to perform
        command: debugging::DebugCommand,
    },
    /// Reply for ServerMessage::GetOrder
    OrderMessage {
        /// Player's order
        order: model::Order,
    },
    /// Signifies finish of the debug update
    DebugUpdateDone {
    },
    /// Request debug state from the app
    RequestDebugState {
    },
}

impl trans::Trans for ClientMessage {
    fn write_to(&self, writer: &mut dyn std::io::Write) -> std::io::Result<()> {
        match self {
            Self::DebugMessage {
                command,
            } => {
                <i32 as trans::Trans>::write_to(&0, writer)?;
                command.write_to(writer)?;
            }
            Self::OrderMessage {
                order,
            } => {
                <i32 as trans::Trans>::write_to(&1, writer)?;
                order.write_to(writer)?;
            }
            Self::DebugUpdateDone {
            } => {
                <i32 as trans::Trans>::write_to(&2, writer)?;
            }
            Self::RequestDebugState {
            } => {
                <i32 as trans::Trans>::write_to(&3, writer)?;
            }
        }
        Ok(())
    }
    fn read_from(reader: &mut dyn std::io::Read) -> std::io::Result<Self> {
        let tag = <i32 as trans::Trans>::read_from(reader)?;
        match tag {
            0 => {
                let command: debugging::DebugCommand = trans::Trans::read_from(reader)?;
                Ok(Self::DebugMessage {
                    command,
                })
            }
            1 => {
                let order: model::Order = trans::Trans::read_from(reader)?;
                Ok(Self::OrderMessage {
                    order,
                })
            }
            2 => {
                Ok(Self::DebugUpdateDone {
                })
            }
            3 => {
                Ok(Self::RequestDebugState {
                })
            }
            _ => Err(std::io::Error::new(
                std::io::ErrorKind::Other,
                format!("Unexpected tag {:?}", tag))),
        }
    }
}